### PR TITLE
Bound incoming HTLC witnessScript to min/max limits

### PR DESCRIPTION
**Pull request by ariard**
Fix a crash where previously we weren't able to detect any accepted
HTLC if its witness-encoded cltv expiry was different from expected
ACCEPTED_HTLC_SCRIPT_WEIGHT. This should work for any cltv expiry
included between 0 and 16777216 on mainnet, testnet and regtest.

### CODE REVIEW

File: lightning/src/ln/channelmonitor.rs [hunk 1]
```diff
@@ -397,6 +396,20 @@ enum InputMaterial {
 	}
 }
 
+#[cfg(test)]
+pub const ACCEPTED_HTLC_SCRIPT_WEIGHT: usize = 138; //Here we have a diff due to HTLC CLTV expiry being < 2^15 in test
```
Comment (TheBlueMatt) [inline lightning/src/ln/channelmonitor.rs]: Hmmm, can we just drop these constants from channelmonitor and replace them with an fn weight_to_type(weight: usize) -> type enum.

Comment (ariard) [reply lightning/src/ln/channelmonitor.rs]: You mean type_to_weight  ? As these constants are used as type assertions based on a known weight

**Comment by TheBlueMatt** [comment]
Well they’re at least used for converting a known weight to the input type. If we want to assert an input has a given weight we can move these constants to functional_test_utils.

**Comment by ariard** [comment]
Updated at 0eb0621, introducing HTLCType::weight_to_htlctype

File: lightning/src/ln/chan_utils.rs [hunk 2]
```diff
@@ -25,6 +25,25 @@ use secp256k1;
...
 
+#[derive(PartialEq)]
+pub(super) enum HTLCType {
+	AcceptedHTLC,
+	OfferedHTLC,
+	NoneHTLC,
+}
+
+impl HTLCType {
+	pub(super) fn weight_to_htlctype(witness_script_len: usize) ->  HTLCType {
```
Comment (TheBlueMatt) [inline lightning/src/ln/chan_utils.rs]: This should return an Option instead of having a NoneHTLC enum

Comment (TheBlueMatt) [inline lightning/src/ln/chan_utils.rs]: Could use some at least basic docs, like "checks if a given tx weight is potentially is pre-signed HTLC transaction"

Comment (ariard) [reply lightning/src/ln/chan_utils.rs]: I disagree a bit here, I don't found it makes callsites clearer, like "else if (if let Some(htlc_type) = weight_to_htlctype(script.len()) { htlc_type == HTLCType::Offered } else { false })"

Comment (TheBlueMatt) [reply lightning/src/ln/chan_utils.rs]: You should be able to make the callsets else if weight_to_htlctype(script.len())  == Some(HTLCType::Offered), which reads fine to me?

Comment (ariard) [reply lightning/src/ln/chan_utils.rs]: My periodic reminder of Rust greatness.

**Comment by ariard** [comment]
Updated at 8c17afd, including #446

File: lightning/src/util/macro_logger.rs [hunk 3]
```diff
@@ -93,15 +94,29 @@ macro_rules! log_route {
...
+					self.0.input[0].witness.len() == 5 {
+				write!(f, "HTLC-success tx")?;
+			} else {
+				for inp in &self.0.input {
+					if !inp.witness.is_empty() {
+						if HTLCType::weight_to_htlctype(inp.witness.last().unwrap().len()) == Some(HTLCType::OfferedHTLC) { write!(f, "preimage-")?; break }
+						else if HTLCType::weight_to_htlctype(inp.witness.last().unwrap().len()) == Some(HTLCType::AcceptedHTLC) { write!(f, "timeout-")?; break }
+					}
+				}
+				write!(f, " tx")?;
```
Comment (TheBlueMatt) [inline lightning/src/util/macro_logger.rs]: nit: awkward space here.

Comment (ariard) [reply lightning/src/util/macro_logger.rs]: Hmm it was intentional at first, but I'm fine too with "preimage-timeout-tx". Corrected

File: lightning/src/ln/chan_utils.rs [hunk 4]
```diff
@@ -25,6 +25,25 @@ use secp256k1;
...
 
+#[derive(PartialEq)]
+pub(crate) enum HTLCType {
+	AcceptedHTLC,
+	OfferedHTLC
+}
+
+impl HTLCType {
+	/// Check if a given tx witnessScript weight matchs one of a pre-signed HTLC
+	pub(crate) fn weight_to_htlctype(witness_script_len: usize) ->  Option<HTLCType> {
```
Comment (TheBlueMatt) [inline lightning/src/ln/chan_utils.rs]: Wait, why is this called "weight", its redeemScript length.

File: lightning/src/ln/chan_utils.rs [hunk 5]
```diff
@@ -25,6 +25,25 @@ use secp256k1;
...
 pub(super) const HTLC_TIMEOUT_TX_WEIGHT: u64 = 663;
 
+#[derive(PartialEq)]
+pub(crate) enum HTLCType {
+	AcceptedHTLC,
+	OfferedHTLC
+}
+
+impl HTLCType {
+	/// Check if a given tx witnessScript weight matchs one of a pre-signed HTLC
```
Comment (TheBlueMatt) [inline lightning/src/ln/chan_utils.rs]: Same here.

File: lightning/src/ln/chan_utils.rs (hunk 4 above)
Comment (ariard) [reply lightning/src/ln/chan_utils.rs]: The spec refers to weight (BOLT3), I always use it at so but yes you're right it's false. Keep witnessScript tho

**Pull request merged**